provider/azure.go

```diff
@@ -0,0 +1,447 @@
+type azureChangeMap map[*dns.Zone][]*endpoint.Endpoint
+
+func (p *AzureProvider) mapChanges(zones []dns.Zone, changes *plan.Changes) (azureChangeMap, azureChangeMap) {
+	ignored := map[string]bool{}
+	deleted := azureChangeMap{}
+	updated := azureChangeMap{}
+
+	mapChange := func(changeMap azureChangeMap, change *endpoint.Endpoint) {
+		zone := p.findZone(zones, change.DNSName)
+		if zone == nil {
+			if _, ok := ignored[change.DNSName]; !ok {
+				ignored[change.DNSName] = true
+				log.Infof("Ignoring changes to '%s' because a suitable Azure DNS zone was not found.", change.DNSName)
+			}
+			return
+		}
+		// Ensure the record type is suitable
+		change.RecordType = suitableType(change)
+		changeMap[zone] = append(changeMap[zone], change)
+	}
+
+	for _, change := range changes.Delete {
+		mapChange(deleted, change)
+	}
+
+	for _, change := range changes.UpdateOld {
+		mapChange(deleted, change)
+	}
+
+	for _, change := range changes.Create {
+		mapChange(updated, change)
+	}
+
+	for _, change := range changes.UpdateNew {
+		mapChange(updated, change)
+	}
+	return deleted, updated
+}
+
+func (p *AzureProvider) findZone(zones []dns.Zone, name string) *dns.Zone {
+	var result *dns.Zone
+
+	// Go through every zone looking for the longest name (i.e. most specific) as a matching suffix
+	for idx := range zones {
+		zone := &zones[idx]
+		if strings.HasSuffix(name, *zone.Name) {
+			if result == nil || len(*zone.Name) > len(*result.Name) {
+				result = zone
+			}
+		}
+	}
+	return result
+}
+
+func (p *AzureProvider) deleteRecords(deleted azureChangeMap) {
+	// Delete records first
+	for zone, endpoints := range deleted {
+		for _, endpoint := range endpoints {
+			name := p.recordSetNameForZone(zone, endpoint)
+			if p.dryRun {
+				log.Infof("Would delete %s record named '%s' for Azure DNS zone '%s'.", endpoint.RecordType, name, *zone.Name)
+			} else {
+				log.Infof("Deleting %s record named '%s' for Azure DNS zone '%s'.", endpoint.RecordType, name, *zone.Name)
+				if _, err := p.recordsClient.Delete(p.resourceGroup, *zone.Name, name, dns.RecordType(endpoint.RecordType), ""); err != nil {
+					log.Errorf(
+						"Failed to delete %s record named '%s' for Azure DNS zone '%s': %v",
+						endpoint.RecordType,
+						name,
+						*zone.Name,
+						err,
+					)
+				}
+			}
+		}
+	}
+}
+
+func (p *AzureProvider) updateRecords(updated azureChangeMap) {
+	for zone, endpoints := range updated {
+		for _, endpoint := range endpoints {
+			name := p.recordSetNameForZone(zone, endpoint)
+			if p.dryRun {
+				log.Infof(
+					"Would update %s record named '%s' to '%s' for Azure DNS zone '%s'.",
+					endpoint.RecordType,
+					name,
+					endpoint.Target,
+					*zone.Name,
+				)
+			} else {
+				log.Infof(
+					"Updating %s record named '%s' to '%s' for Azure DNS zone '%s'.",
+					endpoint.RecordType,
+					name,
+					endpoint.Target,
+					*zone.Name,
+				)
+			}
+			recordSet, err := p.newRecordSet(endpoint)
+			if err == nil && !p.dryRun {
```

Comment:
Could we put this in the `else` branch of the `if p.dryRun` above? Would be more similar to the delete case.